### ISSUE TITLE
Odd behavior on ada files

### ISSUE THREAD

**Issue by glacambre** [vim/vim#2825]
Reproduced on `353eeeaca269ed5e83900bd4a24dc6dc80bb4880`.
When editing the following file:
```ada
case Arg is 
	when Name_Async,
	when Name_Num_Gangs,
end if;
```
with the Ada filetype activated, the following steps:
`ggjVj<C-v>$o$A<Backspace><Esc>`
(with `<C-v>` being Ctrl+v) will insert `case Arg is` at the end of line 3. I don't understand why and think it might be a bug (but I've been wrong about this kind of thing before so please let me know if this is the expected behavior :)).

**Comment by chrisbra** [comment]
I found it a bit hard to follow. Is this the result you are seeing:
```
case Arg is 
	when Name_Async
	when Name_Num_Gangs,case Arg is 
end if;
```
This looks like a bug in the block editing mode.

**Comment by chrisbra** [comment]
BTW: deleting while in block editing mode is not really supported.

**Comment by glacambre** [comment]
Yes, this is exactly what I'm seeing!

Sorry about the steps being hard to follow. Is there another way I could have presented the steps in order to make this issue easier to reproduce?

**Comment by chrisbra** [comment]
So you are trying to remove the last character from the visually selected area. I think a `:s` command would be more beneficial here, `:s/,$//` (in visual mode).

Here is a patch, that should prevent the behaviour that you have been seeing:
```patch
diff --git a/src/ops.c b/src/ops.c
index c59e39a4a..61049741b 100644
--- a/src/ops.c
+++ b/src/ops.c
@@ -2781,7 +2781,10 @@ op_insert(oparg_T *oap, long count1)
         * Subsequent calls to ml_get() flush the firstline data - take a
         * copy of the required string.
         */
-       firstline = ml_get(oap->start.lnum) + bd.textcol;
+       firstline = ml_get(oap->start.lnum);
+       if ((colnr_T)STRLEN(firstline) < bd.textcol)
+           bd.textcol = (colnr_T)STRLEN(firstline);
+       firstline +=bd.textcol;
        if (oap->op_type == OP_APPEND)
            firstline += bd.textlen;
        if (pre_textlen >= 0
```

**Comment by brammool** [comment]
Christian wrote:

> So you are trying to remove the last character from the visually
> selected area. I think a `:s` command would be more beneficial here,
> `:s/,$//` (in visual mode).
> 
> Here is a patch, that should prevent the behaviour that you have been seeing:
> ```patch
> diff --git a/src/ops.c b/src/ops.c
> index c59e39a4a..61049741b 100644
> --- a/src/ops.c
> +++ b/src/ops.c
> @@ -2781,7 +2781,10 @@ op_insert(oparg_T *oap, long count1)
>          * Subsequent calls to ml_get() flush the firstline data - take a
>          * copy of the required string.
>          */
> -       firstline = ml_get(oap->start.lnum) + bd.textcol;
> +       firstline = ml_get(oap->start.lnum);
> +       if ((colnr_T)STRLEN(firstline) < bd.textcol)
> +           bd.textcol = (colnr_T)STRLEN(firstline);
> +       firstline +=bd.textcol;
>         if (oap->op_type == OP_APPEND)
>             firstline += bd.textlen;
>         if (pre_textlen >= 0
> ```

It also adds bd.textlen next.  Maybe that also goes over the end of the
line?  How about this:

	firstline = ml_get(oap->start.lnum) + bd.textcol;
	len = STRLEN(firstline);
	add = bd.textcol;
	if (oap->op_type == OP_APPEND)
	    add += bd.textlen;
	if ((size_t)add > len)
	    firstline += len;  // short line, point to the NUL
	else
	    firstline += add;

We should also have a test for this.

-- 
"Never be afraid to tell the world who you are."
                                        -- Anonymous

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by vim-ml** [comment]
On Mo, 30 Apr 2018, Bram Moolenaar wrote:

> It also adds bd.textlen next.  Maybe that also goes over the end of the
> line?  How about this:
> 
> 	firstline = ml_get(oap->start.lnum) + bd.textcol;
> 	len = STRLEN(firstline);
> 	add = bd.textcol;
> 	if (oap->op_type == OP_APPEND)
> 	    add += bd.textlen;
> 	if ((size_t)add > len)
> 	    firstline += len;  // short line, point to the NUL
> 	else
> 	    firstline += add;
> 
> We should also have a test for this.

How about the attached patch? It includes a test and should fail with 
the current behaviour. I struggled a bit to reproduce the issue with an 
automatic test, but I think I finally got a version that works.

Best,
Christian
-- 
Der Mensch kann nicht leben ohne das sittlich Große, ja, wenn es ihm
entzogen wird, verlangt er danach mit heftigerem Hunger als nach jedem
anderen Ding dieser Erde.
		-- Adalbert Stifter

**Comment by glacambre** [comment]
Thank you for fixing this!
